src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -403,9 +404,23 @@ async def map_module_to_address(
     ):
         return PythonModuleOwners(addresses)
 
+    addresses_by_ancestry = defaultdict(list)
+    for provider in providers:
+        ancestry = calc_ancestry(request.module, provider.addr)
+        addresses_by_ancestry[ancestry].append(provider.addr)
+    minimal_ancestry = min(addresses_by_ancestry.keys())
+    minimal_ancestry_addresses = tuple(addresses_by_ancestry[minimal_ancestry])
+    if len(minimal_ancestry_addresses) == 1:
+        return PythonModuleOwners(minimal_ancestry_addresses)
+
     return PythonModuleOwners((), ambiguous=addresses)
 
 
+def calc_ancestry(module: str, address: Address) -> int:
```

Comment:
So, this was a good idea in theory, but this `calc_ancestry` isn't correct in practice, for two reasons:

1) It's making assumptions about the address that don't hold in general. A provider of a symbol can have anything in its address, it need have no relation to the module name (even if it often does in practice).

2) The `module.count(".")` is the same for all providers, so this does nothing to disambiguate. 

I think we're back to having to compute the ancestry level in `_providers_for_resolve`, which is where we definitely know it, because we know if we looked up the module or its parent.